### PR TITLE
fix(Core/Creatures): Fixed displaying trainer npcflags that are not v…

### LINKED ISSUE #4249: Class/Weapon Trainers incorrect Mouse Icons [$15]
##### CURRENT BEHAVIOUR:
Class trainers offer the same icon for every class, even if they're the incorrect class for that trainer.

In this image, I am a rogue hovering over the warrior trainer - it should just offer a generic gossip icon.

This behavior is present on **every single class trainer in the game.**  This might be in the core and not just the DB - if you re-flag the NPC, you'll end up with the generic gossip icon for same-class trainers.

##### EXPECTED BLIZZLIKE BEHAVIOUR:
Example:

 - This icon should show if you're a hunter talking to the hunter trainer

 - This icon should show if you're a hunter talking to the warrior trainer

##### STEPS TO REPRODUCE THE PROBLEM:
1. Create any class and zone into any starting zone, there are trainers in every starting zone
2. Go to a trainer that isn't for your class
3. Hover your mouse over the NPC and look at the icon

##### AC HASH/COMMIT:
f3f300a1defb

##### OPERATING SYSTEM:
Win10

<bountysource-plugin>

---
There is a **[$15 open bounty](https://www.bountysource.com/issues/95868725-class-weapon-trainers-incorrect-mouse-icons?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
…alid to player.

Fixed #4249.

## Changes Proposed:
-  
-  

## Issues Addressed:
- Closes #4249

## Tests Performed:
- Tested ingame - works okay with class/mount/profession trainers.
-

## Known Issues and TODO List:

- [ ]
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by pklloveyou** [comment]
Tested

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -1033,6 +1033,43 @@ bool Creature::isCanTrainingAndResetTalentsOf(Player* player) const
            && player->getClass() == GetCreatureTemplate()->trainer_class;
 }
 
+bool Creature::IsValidTrainerForPlayer(Player* player, uint32* npcFlags /*= nullptr*/) const
+{
+    if (!IsTrainer())
+        return false;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Creature/Creature.cpp:1039]: please always use brackets, even with 1-liner `if`/`else`/`for`/etc... blocks

File: src/server/game/Entities/Creature/Creature.h [hunk 2]
```diff
@@ -478,6 +478,7 @@ class Creature : public Unit, public GridObject<Creature>, public MovableMapObje
     ///// TODO RENAME THIS!!!!!
     bool isCanInteractWithBattleMaster(Player* player, bool msg) const;
     bool isCanTrainingAndResetTalentsOf(Player* player) const;
+    bool IsValidTrainerForPlayer(Player* player, uint32* npcFlags = nullptr) const;
```
Comment (FrancescoBorzi) [inline src/server/game/Entities/Creature/Creature.h:481]: unless it has side effects, please always mark your methods with `[[nodiscard]]`

**Comment by UltraNix** [comment]
Done.

**Comment by pklloveyou** [comment]
Tested

File: src/server/game/Entities/Creature/Creature.cpp (hunk 1 above)
Comment (FrancescoBorzi) [reply src/server/game/Entities/Creature/Creature.cpp:1040]: @UltraNix still brackets missing in the code below

**Comment by pklloveyou** [comment]
No need to test again

**Comment by UltraNix** [comment]
I don't like Discord.

**Comment by UltraNix** [comment]
Updated.

**Pull request merged**